docs/proposals/gateway-settings.md

```diff
@@ -0,0 +1,176 @@
+spec:
+    controllerName: gateway.nginx.org/nginx-gateway-controller
+    parametersRef:
+        group: gateway.nginx.org/v1alpha1
+        kind: GatewaySettings
+        name: gw-settings
+```
+
+Below is the Golang API for the `GatewaySettings` API:
+
+### Go
+
+```go
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type GatewaySettings struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the GatewaySettings.
+	Spec GatewaySettingsSpec `json:"spec"`
+
+	// Status defines the state of the GatewaySettings.
+	Status GatewaySettingsStatus `json:"status,omitempty"`
+}
+
+type GatewaySettingsSpec struct {
+	// OtelExporter specifies OpenTelemetry export parameters.
+	// +optional
+	OtelExporter *OtelExporter `json:"otelExporter,omitempty"`
+
+	// OtelServiceName is the "service.name" attribute of the Otel resourc.
+	// Default is 'nginx-gateway-fabric:<gateway-name>'.
+	// +optional
+	OtelServiceName *string `json:"otelServiceName,omitempty"`
+
+	// OtelSpanAttributes are custom key/value attributes that are added to each span.
+	// Variables can be included in the values.
+	// +optional
+	OtelSpanAttributes map[string]string `json:"otelSpanAttributes,omitempty"`
+}
+
+type GatewaySettingsStatus struct {
+	// Conditions describe the current conditions of the GatewaySettings.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// OtelExporter specifies OpenTelemetry export parameters.
+type OtelExporter struct {
+	// Interval is the maximum interval between two exports, by default is 5 seconds.
+	// +optional
+	Interval *Duration `json:"interval,omitempty"`
+
+	// BatchSize is the maximum number of spans to be sent in one batch per worker, by default is 512.
+	// +optional
+	BatchSize *int32 `json:"batchSize,omitempty"`
+
+	// BatchCount is the number of pending batches per worker, spans exceeding the limit are dropped,
+	// by default is 4.
+	// +optional
+	BatchCount *int32 `json:"batchCount,omitempty"`
+
+	// Endpoint is the address of OTLP/gRPC endpoint that will accept telemetry data.
+	Endpoint string `json:"endpoint"`
+}
+
+// Duration is a string value representing a duration in time.
+// The format is a subset of the syntax parsed by Golang time.ParseDuration.
+// Examples: 1h, 12m, 30s, 150ms.
+type Duration string
+```
+
+### Status
+
+#### Conditions
+
+The `Accepted` Condition must be populated on the `GatewaySettings` CRD using the reasons defined in the [PolicyCondition API](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/policy_types.go). If these reasons are not sufficient, we can add implementation-specific reasons.
+
+#### GatewayClass Status
+
+While this status does not officially exist, in order to be consistent with other resources with object refs we should add a custom `ResolvedRefs` Condition to the `GatewayClass` when the `parametersRef` is in use.
```

Comment:
Does it make sense to add it to the Gateway API too?

Reply:
Yep, and the Gateway API team agrees.

Reply:
This may be on hold right now. Had further conversation, and while this condition could be useful, the conformance around it needs to be ironed out.

There's also an existing reason, InvalidParameters, that can be used with the Accepted condition. The issue with this, however, is that if we allow dynamic changes to the paramsRef, an invalid config could lead to the GatewayClass being changed from Accepted to not Accepted, and invalidating the entire config tree below it.

Reply:
I feel like we need to allow dynamic changes to the CRD, so a cluster operator doesn't have to disrupt everything just to change the tracing interval. But by doing this, we need to figure out the behavior when the paramsRef is updated to something invalid (or updated to an invalid reference), because if we mark the GatewayClass as not Accepted, that's also very disruptive.

Reply:
Could we use timestamps? Check when the GatewayClass condition was set to False, and check when a Gateway was created?

Reply:
https://github.com/kubernetes-sigs/gateway-api/issues/2911 was created for further discussion

Reply:
> Could we use timestamps? Check when the GatewayClass condition was set to False, and check when a Gateway was created?

I don't think so. 

If we allow NGF to use the last valid version of GatewaySetttings, the following can happen.

Right now if our NGF container restarts, it will try to rebuild the state after the start and it will not see any previous version of valid GatewaySettings. Thus, it will invalidate its Gateway and reflect in the NGINX config of the corresponding container.

However, other replaces will continue running.

That will create different configuration on NGF pods, which will lead for client requests getting different responses (ex. 200 and 404s) depending on which pod the request landed.

Reply:
Ah, I see what you mean.

Reply:
There doesn't seem to be a perfect solution here. We could still Accept the GatewayClass, but set `InvalidParameters` reason on the condition, and simply ignore the GatewaySettings configuration, while allowing everything else to continue to work.

Reply:
let's consider the settings and the consequences of falling to default settings.

Based the current planned settings https://github.com/nginxinc/nginx-gateway-fabric/blob/main/docs/proposals/nginx-extensions.md#gateway-settings

[proxy_protocol](https://nginx.org/en/docs/stream/ngx_stream_proxy_module.html#proxy_protocol) - if this setting was set, falling back to the default value will lead to downtime - NGINX will reject connections from an external load balancer because it will not expect proxy protocol.
[access_log](https://nginx.org/en/docs/http/ngx_http_log_module.html#access_log) - the access log will fall back to the default. not disruptive to client traffic. but probably a nuisance to log analysis. 
[error_log](https://nginx.org/en/docs/ngx_core_module.html#error_log). not disruptive to client traffic. but probably little or no nuisance to log analysis. 
[log_format](https://nginx.org/en/docs/http/ngx_http_log_module.html#log_format) - same as access log.
otel - will not affect client traffic.
[resolver](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver) - will affect client traffic, because any upstreams resolved via DNS will no longer be able to resolve (we don't support this feature currently though)

from this analysis, I don't see a big problem of falling back to default settings if the resource becomes missing and continue Accepting GatewayClass. As long as we don't put any security related settings in GatewaySettings. Because in that case, the configured setting might be rejecting some requests, but the default could suddenly allow them :( 

Also, I think once we split control plane from data plane and start provision data planes, GatewaySettings might include Kubernetes data plane pod specific settings, updates to which will be disruptive - will require rollout of  new pods.

Reply:
It seems that the consequences are either complete removal of the config, or reverting these settings. A case of picking the lesser of two evils.

Reply:
I think the lesser evil is reverting to the default settings. I think we should be loud about it, though. Log an error, write an event, and add a condition. 

> As long as we don't put any security related settings in GatewaySettings. Because in that case, the configured setting might be rejecting some requests, but the default could suddenly allow them :(

Agreed. If we go with this approach, we should mention this restriction in the proposal.

Reply:
For now I'll write up this approach, but I'm going to chat with the Gateway API team next week about this and see if they have any other thoughts.

Reply:
https://github.com/kubernetes-sigs/gateway-api/issues/2911 exists to track discussion on these couple of issues. I'll give it a few days for feedback, but if nothing clear comes through, then we can start with the proposed approach. Can always come back and change it before our next release.

Reply:
It's possible that API will be rejected. It's been rejected by the k8s API group before. So definitely not worth waiting for

Reply:
I wasn't planning on waiting for any actual API changes to be made upstream. More-so just getting opinions, and we can implement first before switching to the updated API.

Reply:
Makes sense. I just wanted to mention that the proposed API may not be accepted.